### ISSUE TITLE
having SPD log as DLL

### ISSUE THREAD

**Issue by naveenpinto** [gabime/spdlog#756]
Hi 
  has anyone done SPD log as a DLL in there project. as we are using template function and variable arguments i am not able to do a DLL export of these function, can you please let me know what is the best what is the best way to make it as a shared lib????

**Comment by gabime** [comment]
spdlog is header only lib.. Let your users include spdlog headers too, and export to them the logger instances.

**Comment by naveenpinto** [comment]
as the trace, debug.. etc function are template Variadic we are not able to export them as a DLL, nor write a wrapper as we have to pass variadic template arguments, we are able to use them as a static library, this is fine for a single process system, but for a multi process system we can use a static lib, we require a DLL. 
Any thought on this.

**Comment by rpopescu** [comment]
"but for a multi process system we can't use a static lib, we require a DLL" - is that really so? why?

**Comment by gabime** [comment]
Is your objective is multiple processes logging to same file? spdlog doesnt support this. Each process should write to his own file.

**Comment by naveenpinto** [comment]
yes but if we can spd log as a DLL then all process can write to the same logs, 
also it will be good if we can have spdlog as a DLL, any though on how we can have it as a dll cause the bottle neck is template and variadics :), which is making me use it as a static lib( using a wrapper).

**Comment by gabime** [comment]
> yes but if we can spd log as a DLL then all process can write to the same logs,

Won't work.. I suggest appending pid to name of the file, so each process writes to its own file

**Comment by arthur-tacca** [comment]
There is a case where it would actually be useful for spdlog to be in a DLL: where you have multiple DLLs within a *single* process and you want those DLLs to call spdlog functions. I'm not sure if that's what the OP meant or not.

In that case I think you would want the logging functions to be inline in the client DLLs and the sink writing functions to be in spdlog's DLL, passing log record objects between the two. It would be doable but a lot of work and the code would have to be awkwardly reorganised.

The usual alternative, which doesn't require complicating spdlog's code, is that users wanting this functionality write their own wrapper with a C interface, exactly as OP said. IMO this seems like a better trade off given the number of people likely to be in this situation.

**Comment by naveenpinto** [comment]
Hi i checked the cstyle variadics but it seems to be quite flaky and error prone, also we can use the new fmt standard which is really nice to log objects, does any one tried c style interface for SPD log library, please share sample code for this.

**Comment by Wambou** [comment]
Hi,
Facing the same problem as described by @arthur-tacca and I wanted to give the solution I found to circumvent the problem.

My DLLs have a function call "setup_logger" which construct the logger for the DLL, with it's own name and using sinks from the main process. This function returns the created logger that I can register in the main program.

By doing that, and from the small set of tests I did so far, I can change log level using spdlog::set_level() that will also influence the logger inside the DLL.

I hope this can help some users and maybe that could be added into the documentation in a special section.

**Comment by gabime** [comment]
@Wambou Thanks!
Please note that spdlog now also provide a logger->clone() method which can simplify your wiki example

**Comment by Wambou** [comment]
Thank you for the information.
I still use spdlog 0.16.3, hence the "old style".
